### PR TITLE
64bit indexing fused adam

### PR DESCRIPTION
**Pull request by garrett4wade**
## The Issue

Applying `FusedAdam` on large tensors will cause an error `CUDA error: an illegal memory access was encountered`.

https://github.com/microsoft/DeepSpeed/issues/3429

https://github.com/NVIDIA/apex/issues/1654

## PR Content

Following the solution in the apex repository (https://github.com/NVIDIA/apex/pull/1765), changing indexing type to `int64` if necessary.

### CODE REVIEW

**Comment by garrett4wade** [comment]
> @garrett4wade please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Review by mrwyattii — APPROVED** [review]

File: csrc/adam/multi_tensor_adam.cu [hunk 1]
```diff
@@ -30,7 +30,7 @@ typedef enum : int {
 
 using MATH_T = float;
 
-template <typename T>
+template <typename T, typename index_t>
 struct AdamFunctor {
     __device__ __forceinline__ void operator()(int chunk_size,
```
Comment (loadams) [inline csrc/adam/multi_tensor_adam.cu:35]: FYI this chunk_size is still an int and not int64/index_t

**Review by loadams — APPROVED** [review]

File: csrc/adam/multi_tensor_adam.cu (hunk 1 above)
Comment (loadams) [reply csrc/adam/multi_tensor_adam.cu:35]: @garrett4wade - thoughts on this?

**Comment by loadams** [comment]
FYI @garrett4wade - thoughts on the comments above or merging this PR?

**Pull request merged**